### PR TITLE
[CGS] Add features and improvements

### PR DESCRIPTION
**Pull request by amontoison**
With the help of BiCGSTAB references, I understand a little bit more the CGS method. I updated the code to add some functionalities:
- CGS is based on the Lanczos biorthogonalization process and the user can now provide his own second vector `c`. It could be also easier to compare with BiLQ or QMR.
- Use the `N` parameter instead of `M` for a right preconditioner. It allows us to stay consistent with the other Krylov methods.
- Add the possibility to use a left preconditioner `M`.
- Add an example to show how an ILU factorization can be used as a split preconditioner.

### CODE REVIEW

**Comment by dpo** [comment]
I think in CGS we use c = r implicitly to avoid the user providing a vector c such that rho = 0.

**Comment by amontoison** [comment]
c is not argument but a keyword argument. By default, `r̅₀=c=b` such that `rho  ≠ 0`.

File: examples/test_cgs.jl [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+A = spdiagm(
+      -1 => fill(-1.0, n - 1), 
+       0 => fill(3.0, n), 
+       1 => fill(-2.0, n - 1)
+    )
+Id = sparse(1.0I, n, n)
+A = kron(A, Id) + kron(Id, A)
+A = kron(A, Id) + kron(Id, A)
+x = ones(N)
+b = A * x
```
Comment (amontoison) [inline examples/test_cgs.jl:15]: After #221, I will replace these lines by
```julia
include("../test/test_utils.jl")
A, b = kron_unsymmetric()
```

Comment (dpo) [reply examples/test_cgs.jl:15]: I merged #221.

**Review by dpo — APPROVED** [review]

**Comment by dpo** [comment]
Thanks.

**Pull request merged**